Document/0x05e-Testing-Cryptography.md

```diff
@@ -6,23 +6,126 @@
 
--- TODO [Clarify the purpose of "Use the &lt;sup&gt; tag to reference external sources, e.g. Meyer's recipe for tomato soup<sup>[1]</sup>."] --
+Android SDK provides mechanisms for specifying secure key generation and use. Android 6.0 (Marshmallow, API 23) introduced the `KeyGenParameterSpec` class that can be used to ensure the correct key usage in the application. 
 
--- TODO [Develop Static Analysis with source code of "Verifying the Configuration of Cryptographic Standard Algorithms"] --
+Here's an example of using AES/CBC/PKCS7Padding on API 23+:
 
-#### Dynamic Analysis
+```
+String keyAlias = "MySecretKey";
+
+KeyGenParameterSpec keyGenParameterSpec = new KeyGenParameterSpec.Builder(keyAlias,
+        KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+        .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+        .setRandomizedEncryptionRequired(true)
+        .build();
+
+KeyGenerator keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES,
+        "AndroidKeyStore");
+keyGenerator.init(keyGenParameterSpec);
+
+SecretKey secretKey = keyGenerator.generateKey();
+```
+
+The `KeyGenParameterSpec` indicates that the key can be used for encryption and decryption, but not for other purposes, such as signing or verifying. It further specifies the block mode (CBC), padding (PKCS7), and explicitly specifies that randomized encryption is required (this is the default.) `"AndroidKeyStore"` is the name of the cryptographic service provider used in this example.
+
+GCM is another AES block mode that provides additional security benefits over other, older modes. In addition to being cryptographically more secure, it also provides authentication. When using CBC (and other modes), authentication would need to be performed separately, using HMACs (see the Reverse Engineering chapter). Note that GCM is the only mode of AES that does not support paddings.<sup>[3], [5]</sup>
+
+Attempting to use the generated key in violation of the above spec would result in a security exception.
+
+Here's an example of using that key to decrypt:
+
+```
+String AES_MODE = KeyProperties.KEY_ALGORITHM_AES
+        + "/" + KeyProperties.BLOCK_MODE_CBC
+        + "/" + KeyProperties.ENCRYPTION_PADDING_PKCS7;
+KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+
+// byte[] input
+Key key = keyStore.getKey(keyAlias, null);
+
+Cipher cipher = Cipher.getInstance(AES_MODE);
+cipher.init(Cipher.ENCRYPT_MODE, key);
+
+byte[] encryptedBytes = cipher.doFinal(input);
+byte[] iv = cipher.getIV();
+```
+
+Here's how that cipher text would be decrypted:
+
+```
+// byte[] input
+// byte[] iv
+Key key = keyStore.getKey(AES_KEY_ALIAS, null);
+
+Cipher cipher = Cipher.getInstance(AES_MODE);
+IvParameterSpec params = new IvParameterSpec(iv);
```

Comment:
Added a note that IV needs to be saved alongside the cipher text. Didn't provide details on how to save them, since they could be either stored separately or concatenated and split.